A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,164 @@
+each endpoint on the ring will be extracted from a pre-defined name resolver
+attribute called `hash_key`. If this attribute is set, then the endpoint is
+placed on the ring by hashing its value. If this attribute is not set, then the
+endpoint IP address is used (current behavior). The location of an existing
+endpoint on the ring changes if its `hash_key` name resolver attribute changes.
+
+The xDS resolver will be changed so that when converting EDS responses to
+resolver endpoints, it will set the `hash_key` name resolver attribute to the
+value of
+[LbEndpoint.Metadata](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#envoy-v3-api-field-config-endpoint-v3-lbendpoint-metadata)
+`envoy.lb` `hash_key` field, as described in [Envoy's documentation for the ring
+hash load
+balancer](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#ring-hash).
+
+### LB Policy Config Changes
+
+After the addition of this field, the `ring_hash` LB policy config will be:
+
+    message RingHashLoadBalancingConfig {
+      // A client-side cap will limit these values.  If either of these values
+      // are greater than the client-side cap, they will be treated as the
+      // client-side cap.  The default cap is 4096.
+      uint64 min_ring_size = 1;        // Optional, defaults to 1024.
+      uint64 max_ring_size = 2;        // Optional, defaults to 4096, max is 8M.
+
+      string request_metadata_key = 3; // Optional, defaults to the empty string.
+    }
+
+
+### Temporary environment variable protection
+
+Explicitly setting the request hash key cannot possibly lead to problem with
+existing deployment because the new behavior requires setting a load balancing
+policy configuration field that did not exist before. Therefore, it is not gated
+behind an environment variable.
+
+The second behavior change will be enabled by the
+`GRPC_EXPERIMENTAL_XDS_RING_HASH_ENDPOINT_HASH_KEY` environment variable. This
+will protect from the case where an xDS control plane is already setting the
+`LbEndpoint.Metadata` `envoy.lb` `hash_key` field, in which case deploying this
+new behavior would churn all endpoint hash keys, which could lead to
+problems. This environment variable will be removed once the feature has proven
+stable.
+
+## Rationale
+
+We originally proposed using language specific interfaces to set the request
+hash key. The advantage would have been that the request hash key would not have
+to be exposed through gRPC metadata. However, this would have required defining
+language specific APIs, which would increase the complexity of this change.
+ 
+We also discussed the option of exposing all `LbEndpoint.metadata` from EDS
+through name resolver attributes, instead of only extracting the specific
+`hash_key` attribute, so as to make them available to custom LB policies. We
+decided to keep only extract `hash_key` to limit the scope of this gRFC.
```

Comment:
How this decision was made? IMO it is the same amount of code to copy all lbEndpoint.metadata to attributes as just process `hash_key`

Reply:
Originally I didn't do that because I thought that would cause more discussions, but in principle I agree with you.

You should have a look at https://github.com/grpc/grpc/issues/33356#issuecomment-1895299488 and response from Mark, it's very interesting.